tests/integration/test_backups.py

```diff
@@ -0,0 +1,332 @@
+        backup_id = result.results["backup-id"]
+
+        # list backups again and ensure new backup id exists
+        logger.info("Listing backup ids post backup")
+
+        action = await zeroth_unit.run_action(action_name="list-backups")
+        result = await action.wait()
+        new_backup_ids = json.loads(result.results["backup-ids"])
+
+        assert sorted(new_backup_ids) == sorted(backup_ids + [backup_id])
+
+        backups_by_cloud[cloud_name] = backup_id
+
+    # insert data into cluster after backup
+    logger.info("Inserting value after backup")
+    value_after_backup = await insert_data_into_mysql_and_validate_replication(
+        ops_test,
+        DATABASE_NAME,
+        TABLE_NAME,
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_restore_on_same_cluster(ops_test: OpsTest, mysql_charm_series: str) -> None:
+    """Test to restore a backup to the same mysql cluster."""
+    mysql_application_name = await deploy_and_scale_mysql(ops_test, mysql_charm_series)
+
+    logger.info("Scaling mysql application to 1 unit")
+    async with ops_test.fast_forward():
+        await scale_application(ops_test, mysql_application_name, 1)
+
+    mysql_unit = ops_test.model.units[f"{mysql_application_name}/0"]
+    mysql_unit_address = await get_unit_ip(ops_test, mysql_unit.name)
+    server_config_credentials = await get_server_config_credentials(mysql_unit)
+
+    for cloud_name, config in CLOUD_CONFIGS.items():
+        assert backups_by_cloud[cloud_name]
+
+        # set the s3 config and credentials
+        logger.info(f"Syncing credentials for {cloud_name}")
+
+        await ops_test.model.applications[S3_INTEGRATOR].set_config(config)
+        action = await ops_test.model.units[f"{S3_INTEGRATOR}/0"].run_action(
+            "sync-s3-credentials",
+            **CLOUD_CREDENTIALS[cloud_name],
+        )
+        await action.wait()
+
+        await ops_test.model.wait_for_idle(
+            apps=[mysql_application_name, S3_INTEGRATOR],
+            status="active",
+            timeout=TIMEOUT,
+        )
+
+        # restore the backup
+        logger.info(f"Restoring backup with id {backups_by_cloud[cloud_name]}")
+
+        action = await mysql_unit.run_action(
+            action_name="restore", **{"backup-id": backups_by_cloud[cloud_name]}
+        )
+        result = await action.wait()
+        assert result.results.get("Code") == "0"
+
+        # ensure the correct inserted values exist
+        logger.info(
+            "Ensuring that the pre-backup inserted value exists in database, while post-backup inserted value does not"
+        )
+        select_values_sql = [f"SELECT id FROM `{DATABASE_NAME}`.`{TABLE_NAME}`"]
+
+        values = await execute_queries_on_unit(
+            mysql_unit_address,
+            server_config_credentials["username"],
+            server_config_credentials["password"],
+            select_values_sql,
+        )
+        assert values == [value_before_backup]
```

Comment:
please broke data before run restoration for another cloud
```
    logger.info("Inserting value after backup")
    value_after_backup = await insert_data_into_mysql_and_validate_replication(
        ops_test,
        DATABASE_NAME,
        TABLE_NAME,
    )
```

Reply:
addressed in [f5d8606](https://github.com/canonical/mysql-operator/pull/145/commits/f5d8606cd2ce7237bf3051ba1f9b81ea6b2c8a7c)